File: iOS/DataBet/Sources/DataBet/Domain/PoolGamblerBetDummyProvider.swift
import Foundation
import Core

private let countries: [(String, String)] = [
    ("ar", "Argentina"),
    ("at", "Austria"),
    ("au", "Australia"),
    ("ba", "Bosnia and Herzegovina"),
    ("be", "Belgium"),
    ("br", "Brazil"),
    ("ca", "Canada"),
    ("cd", "Congo"),
    ("ch", "Switzerland"),
    ("ci", "Ivory Coast"),
    ("co", "Colombia"),
    ("cv", "Cape Verde"),
    ("cw", "Curaçao"),
    ("cz", "Czech Republic"),
    ("de", "Germany"),
    ("dz", "Algeria"),
    ("ec", "Ecuador"),
    ("eg", "Egypt"),
    ("es", "Spain"),
    ("fr", "France"),
    ("gb_eng", "England"),
    ("gb_sct", "Scotland"),
    ("gh", "Ghana"),
    ("hr", "Croatia"),
    ("ht", "Haiti"),
    ("iq", "Iraq"),
    ("ir", "Iran"),
    ("jo", "Jordan"),
    ("jp", "Japan"),
    ("kr", "South Korea"),
    ("ma", "Morocco"),
    ("mx", "Mexico"),
    ("nl", "Netherlands"),
    ("no", "Norway"),
    ("nz", "New Zealand"),
    ("pa", "Panama"),
    ("pt", "Portugal"),
    ("py", "Paraguay"),
    ("qa", "Qatar"),
    ("sa", "Saudi Arabia"),
    ("se", "Sweden"),
    ("sn", "Senegal"),
    ("tn", "Tunisia"),
    ("tr", "Turkey"),
    ("us", "USA"),
    ("uy", "Uruguay"),
    ("uz", "Uzbekistan"),
    ("za", "South Africa"),
]

func dummyPoolGamblerBet() -> PoolGamblerBet {
    let homeCountry = countries.randomElement()!
    var awayCountry = countries.randomElement()!
    while awayCountry == homeCountry {
        awayCountry = countries.randomElement()!
    }
    let isComputed = Bool.random()
    let isLocked = isComputed || Bool.random()

    return PoolGamblerBet(
        poolId: "pool123",
        gamblerId: "gambler456",
        gamblerUsername: "gambler456@example.com",
        matchId: "match789",
        homeTeamId: homeCountry.0,
        homeTeamName: homeCountry.1,
        awayTeamId: awayCountry.0,
        awayTeamName: awayCountry.1,
        matchScore: isComputed ? TeamScore(homeTeamValue: Int.random(in: 0...5), awayTeamValue: Int.random(in: 0...5)) : nil,
        betScore: TeamScore(homeTeamValue: Int.random(in: 0...5), awayTeamValue: Int.random(in: 0...5)),
        score: isComputed ? Int.random(in: 0...30) : nil,
        matchDateTime: Date(),
        isLocked: isLocked,
        isComputed: isComputed
    )
}


func poolGamblerBetDummyModels() -> [PoolGamblerBet] {
    (1...10).map { i in
        let homeCountry = countries.randomElement()!
        var awayCountry = countries.randomElement()!
        while awayCountry == homeCountry {
            awayCountry = countries.randomElement()!
        }
        let isComputed = Bool.random()
        let isLocked = isComputed || Bool.random()

        return PoolGamblerBet(
            poolId: "pool\(Int.random(in: 100...999))",
            gamblerId: "gambler\(Int.random(in: 100...999))",
            gamblerUsername: "gambler\(i)@example.com",
            matchId: "match\(Int.random(in: 100...999))",
            homeTeamId: homeCountry.0,
            homeTeamName: homeCountry.1,
            awayTeamId: awayCountry.0,
            awayTeamName: awayCountry.1,
            matchScore: isComputed ? TeamScore(homeTeamValue: Int.random(in: 0...5), awayTeamValue: Int.random(in: 0...5)) : nil,
            betScore: TeamScore(homeTeamValue: Int.random(in: 0...5), awayTeamValue: Int.random(in: 0...5)),
            score: Int.random(in: 0...10),
            matchDateTime: Date(),
            isLocked: isLocked,
            isComputed: isComputed
        )
    }
}

func poolGamblerBetPendingDummyModels() -> [PoolGamblerBet] {
    (1...10).map { i in
        let homeCountry = countries.randomElement()!
        var awayCountry = countries.randomElement()!
        while awayCountry == homeCountry {
            awayCountry = countries.randomElement()!
        }
        let isLocked = Bool.random()

        return PoolGamblerBet(
            poolId: "pool\(Int.random(in: 100...999))",
            gamblerId: "gambler\(Int.random(in: 100...999))",
            gamblerUsername: "gambler\(i)@example.com",
            matchId: "match\(Int.random(in: 100...999))",
            homeTeamId: homeCountry.0,
            homeTeamName: homeCountry.1,
            awayTeamId: awayCountry.0,
            awayTeamName: awayCountry.1,
            matchScore: nil,
            betScore: TeamScore(homeTeamValue: Int.random(in: 0...5), awayTeamValue: Int.random(in: 0...5)),
            score: Int.random(in: 0...10),
            matchDateTime: Date().addingTimeInterval(-30 * 60),
            isLocked: isLocked,
            isComputed: false
        )
    }
}

func poolGamblerBetFinishedDummyModels() -> [PoolGamblerBet] {
    (1...10).map { i in
        let homeCountry = countries.randomElement()!
        var awayCountry = countries.randomElement()!
        while awayCountry == homeCountry {
            awayCountry = countries.randomElement()!
        }

        return PoolGamblerBet(
            poolId: "pool\(Int.random(in: 100...999))",
            gamblerId: "gambler\(Int.random(in: 100...999))",
            gamblerUsername: "gambler\(i)@example.com",
            matchId: "match\(Int.random(in: 100...999))",
            homeTeamId: homeCountry.0,
            homeTeamName: homeCountry.1,
            awayTeamId: awayCountry.0,
            awayTeamName: awayCountry.1,
            matchScore: TeamScore(homeTeamValue: Int.random(in: 0...5), awayTeamValue: Int.random(in: 0...5)),
            betScore: TeamScore(homeTeamValue: Int.random(in: 0...5), awayTeamValue: Int.random(in: 0...5)),
            score: Int.random(in: 0...10),
            matchDateTime: Date(),
            isLocked: true,
            isComputed: true
        )
    }
}

File: Android/bet/src/main/java/com/felipearpa/tyche/bet/PoolGamblerBetDummyProvider.kt
package com.felipearpa.tyche.bet

import com.felipearpa.tyche.core.type.TeamScore
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import kotlin.time.Clock
import kotlin.time.ExperimentalTime

private val countries = listOf(
    "ar" to "Argentina",
    "at" to "Austria",
    "au" to "Australia",
    "ba" to "Bosnia and Herzegovina",
    "be" to "Belgium",
    "br" to "Brazil",
    "ca" to "Canada",
    "cd" to "Congo",
    "ch" to "Switzerland",
    "ci" to "Ivory Coast",
    "co" to "Colombia",
    "cv" to "Cape Verde",
    "cw" to "Curaçao",
    "cz" to "Czech Republic",
    "de" to "Germany",
    "dz" to "Algeria",
    "ec" to "Ecuador",
    "eg" to "Egypt",
    "es" to "Spain",
    "fr" to "France",
    "gb_eng" to "England",
    "gb_sct" to "Scotland",
    "gh" to "Ghana",
    "hr" to "Croatia",
    "ht" to "Haiti",
    "iq" to "Iraq",
    "ir" to "Iran",
    "jo" to "Jordan",
    "jp" to "Japan",
    "kr" to "South Korea",
    "ma" to "Morocco",
    "mx" to "Mexico",
    "nl" to "Netherlands",
    "no" to "Norway",
    "nz" to "New Zealand",
    "pa" to "Panama",
    "pt" to "Portugal",
    "py" to "Paraguay",
    "qa" to "Qatar",
    "sa" to "Saudi Arabia",
    "se" to "Sweden",
    "sn" to "Senegal",
    "tn" to "Tunisia",
    "tr" to "Turkey",
    "us" to "USA",
    "uy" to "Uruguay",
    "uz" to "Uzbekistan",
    "za" to "South Africa",
)

@OptIn(ExperimentalTime::class)
fun poolGamblerBetFakeModel() =
    PoolGamblerBetModel(
        poolId = "X".repeat(15),
        gamblerId = "X".repeat(15),
        gamblerUsername = "X".repeat(25),
        matchId = "X".repeat(15),
        homeTeamId = "co",
        homeTeamName = "X".repeat(25),
        awayTeamId = "br",
        awayTeamName = "X".repeat(25),
        matchScore = TeamScore(100, 100),
        betScore = TeamScore(100, 100),
        score = 10,
        matchDateTime = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()),
        isLocked = true,
        isComputed = false,
    )

@OptIn(ExperimentalTime::class)
fun poolGamblerBetDummyModel(): PoolGamblerBetModel {
    val homeCountry = countries.random()
    var awayCountry = countries.random()
    while (awayCountry == homeCountry) {
        awayCountry = countries.random()
    }
    val isComputed = (0..1).random() == 1
    val isLocked = isComputed || (0..1).random() == 1

    return PoolGamblerBetModel(
        poolId = "pool123",
        gamblerId = "gambler456",
        gamblerUsername = "gambler456@example.com",
        matchId = "match789",
        homeTeamId = homeCountry.first,
        homeTeamName = homeCountry.second,
        awayTeamId = awayCountry.first,
        awayTeamName = awayCountry.second,
        matchScore = if (isComputed) TeamScore((0..5).random(), (0..5).random()) else null,
        betScore = TeamScore((0..5).random(), (0..5).random()),
        score = if (isComputed) (0..30).random() else null,
        matchDateTime = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()),
        isLocked = isLocked,
        isComputed = isComputed,
    )
}

@OptIn(ExperimentalTime::class)
fun poolGamblerBetDummyModels(): List<PoolGamblerBetModel> =
    (1..10).map { i ->
        val homeCountry = countries.random()
        var awayCountry = countries.random()
        while (awayCountry == homeCountry) {
            awayCountry = countries.random()
        }
        val isComputed = (0..1).random() == 1
        val isLocked = isComputed || (0..1).random() == 1

        PoolGamblerBetModel(
            poolId = "pool${(100..999).random()}",
            gamblerId = "gambler${(100..999).random()}",
            gamblerUsername = "gambler${i}@example.com",
            matchId = "match${(100..999).random()}",
            homeTeamId = homeCountry.first,
            homeTeamName = homeCountry.second,
            awayTeamId = awayCountry.first,
            awayTeamName = awayCountry.second,
            matchScore = if (isComputed) TeamScore((0..5).random(), (0..5).random()) else null,
            betScore = TeamScore((0..5).random(), (0..5).random()),
            score = (0..10).random(),
            matchDateTime = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()),
            isLocked = isLocked,
            isComputed = isComputed,
        )
    }

@OptIn(ExperimentalTime::class)
fun poolGamblerBetFinishedDummyModels(): List<PoolGamblerBetModel> =
    (1..10).map { i ->
        val homeCountry = countries.random()
        var awayCountry = countries.random()
        while (awayCountry == homeCountry) {
            awayCountry = countries.random()
        }

        PoolGamblerBetModel(
            poolId = "pool${(100..999).random()}",
            gamblerId = "gambler${(100..999).random()}",
            gamblerUsername = "gambler${i}@example.com",
            matchId = "match${(100..999).random()}",
            homeTeamId = homeCountry.first,
            homeTeamName = homeCountry.second,
            awayTeamId = awayCountry.first,
            awayTeamName = awayCountry.second,
            matchScore = TeamScore((0..5).random(), (0..5).random()),
            betScore = TeamScore((0..5).random(), (0..5).random()),
            score = (0..10).random(),
            matchDateTime = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()),
            isLocked = true,
            isComputed = true,
        )
    }

@OptIn(ExperimentalTime::class)
fun poolGamblerBetPendingDummyModels(): List<PoolGamblerBetModel> =
    (1..10).map { i ->
        val homeCountry = countries.random()
        var awayCountry = countries.random()
        while (awayCountry == homeCountry) {
            awayCountry = countries.random()
        }
        val isLocked = (0..1).random() == 1

        PoolGamblerBetModel(
            poolId = "pool${(100..999).random()}",
            gamblerId = "gambler${(100..999).random()}",
            gamblerUsername = "gambler${i}@example.com",
            matchId = "match${(100..999).random()}",
            homeTeamId = homeCountry.first,
            homeTeamName = homeCountry.second,
            awayTeamId = awayCountry.first,
            awayTeamName = awayCountry.second,
            matchScore = null,
            betScore = TeamScore((0..5).random(), (0..5).random()),
            score = (0..10).random(),
            matchDateTime = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()),
            isLocked = isLocked,
            isComputed = false,
        )
    }
